core/src/main/scala/kafka/log/Log.scala

```diff
@@ -609,24 +716,22 @@ class Log(val dir: File,
    * logSegment will be rolled if one of the following conditions met
    * <ol>
    * <li> The logSegment is full
-   * <li> The maxTime has elapsed
+   * <li> The maxTime has elapsed since the largest timestamp in the segment (or since the create time if there is no
+   * message having a timestamp)
    * <li> The index is full
    * </ol>
    * @return The currently active segment after (perhaps) rolling to a new segment
    */
   private def maybeRoll(messagesSize: Int): LogSegment = {
     val segment = activeSegment
+    val reachedRollMs = segment.timeWaitedForRoll(time.milliseconds) > config.segmentMs - segment.rollJitterMs
```

Comment:
In the proposal of KIP-33, we changed the behavior of log rolling to be that if the largestTimestamp in the segment hasn't changed for the log rolling time, we will roll another log segment. Thinking about this a bit more, I am wondering if it's better to preserve the current log rolling logic based on the log segment create time. The current behavior provides a nice property that a log segment is guaranteed to be rolled within certain amount of time. This is useful since there are use cases where people may have some sensitive data that has to be deleted or cleaned after certain amount of time. The new behavior will make this harder to enforce since the active segment is never deleted or cleaned and there is no time bound on when the new segment can be rolled.

The log deletion policy can still be based on the timestamp in the message.